chore: update to electron 9-x-y

deepak1556 opened the pull request:
Fixes https://github.com/microsoft/vscode/issues/100740
Fixes https://github.com/microsoft/vscode/issues/98400
Fixes https://github.com/microsoft/vscode/issues/101069
Fixes https://github.com/microsoft/vscode/issues/88873
Fixes https://github.com/microsoft/vscode/issues/78867

Also fixes https://github.com/microsoft/vscode/issues?q=is%3Aopen+is%3Aissue+label%3Afixed-in-electron-8

**TODO**

- [x] Fix disabling crash reporter via argv.json https://github.com/microsoft/vscode/issues/97123

deepak1556 commented on src/main.js:
```diff
@@ -55,22 +59,30 @@ if (crashReporterDirectory) {
+	const isMacintosh = (process.platform === 'darwin');
+	const isLinux = (process.platform === 'linux');
+	submitURL = isWindows ? appCenter[process.arch === 'ia32' ? 'win32-ia32' : 'win32-x64'] : isLinux ? appCenter[`linux-x64`] : appCenter.darwin;
+	//submitURL = submitURL.concat('&uid=', crashReporterId, '&iid=', crashReporterId, '&sid=', info.sessionId);
```
how can we get `ITelemetryInfo.sessionId` here ?

bpasero replied:
@deepak1556 not just that, but also the `crashReporterId`, no? My understanding is:
* `crashReporterId` is a unique number per installation (e.g. we store it in global storage only once)
* `sessionId` is a unique number that changes each time you start VSCode

The latter is probably used to be able to correlate a crash with telemetry session, but at least I never ever used this for investigating crashes. Maybe it would be sufficient to simply generate another UUID here if it is needed.

The crash reporter ID is more interesting... here we do not have access to the storage service yet. Maybe it could be written into `argv.json`? Since we already need to read that file on startup, we can reduce the perf hit this has.

deepak1556 replied:
Thinking a bit more, encoding both these metadata in the url makes it difficult to have them on launch but there is way to provide additional metadata from each process even after launch using [crashReporter.addExtraParameter](https://github.com/electron/electron/blob/9-x-y/docs/api/crash-reporter.md#crashreporteraddextraparameterkey-value) , I just to need to refactor the appcenter implementation to recognize that.

deepak1556 replied:
Only work left is on the backend, will create a PR today for it in the electron repo.

bpasero commented on build/azure-pipelines/product-build.yml:
```diff
@@ -156,8 +156,8 @@ trigger: none
 schedules:
-- cron: "0 5 * * Mon-Fri"
-  displayName: Mon-Fri at 7:00
+- cron: "0 6 * * Mon-Fri"
```
@deepak1556 this needs to be excluded before merging

bpasero commented on src/bootstrap-fork.js:
```diff
@@ -172,18 +169,4 @@ function terminateWhenParentTerminates() {
 	}
 }
 
-function configureCrashReporter() {
```
@deepak1556 will Electron 9 enable the crash reporter for each child process automatically? How in AppCenter would we be able to tell apart where the crash is coming from?

deepak1556 replied:
Yup it will automatically watch all the child processes including the ones launched by node fork.

> How in AppCenter would we be able to tell apart where the crash is coming from?

we haven't been distinguishing them currently in appcenter, am I missing something here ?

bpasero replied:
Yes you are right, though it would be helpful if there was a way I guess. I think we can already tell apart renderer from main process, no? Would be great to see the executable name appear as metadata somehow.

deepak1556 replied:
we are sending the executable name currently but appcenter isn't displaying them in anyway :(

deepak1556 replied:
A little correction, for linux we need to explicitly start the crash reporter for child processes because of breakpad, have accommodated for it.

bpasero commented on src/vs/workbench/contrib/welcome/telemetryOptOut/browser/telemetryOptOut.ts:
```diff
@@ -139,7 +139,6 @@ export abstract class AbstractTelemetryOptOut implements IWorkbenchContribution
 						run: () => {
 							logTelemetry(true);
 							this.configurationService.updateValue('telemetry.enableTelemetry', false);
-							this.configurationService.updateValue('telemetry.enableCrashReporter', false);
```
@deepak1556 we should write to `argv.json` in this case when the user wants to opt out

bpasero commented on src/main.js:
```diff
@@ -55,22 +59,30 @@ if (crashReporterDirectory) {
+}
+
+// Start crash reporter for all processes
+const productName = (product.crashReporter && product.crashReporter.productName) || product.nameShort;
```
@deepak1556 I think by now you can write this and the next line as `product.crashReporter?.productName) || product.nameShort` (optional chaining).

bpasero commented on src/main.js:
```diff
@@ -55,22 +59,30 @@ if (crashReporterDirectory) {
+	submitURL = isWindows ? appCenter[process.arch === 'ia32' ? 'win32-ia32' : 'win32-x64'] : isLinux ? appCenter[`linux-x64`] : appCenter.darwin;
+	//submitURL = submitURL.concat('&uid=', crashReporterId, '&iid=', crashReporterId, '&sid=', info.sessionId);
 }
+crashReporter.start({
```
Besides checking for `argv.json` disablement, we also still need to support `disable-crash-reporter` command line argument for disablement.

bpasero commented on src/main.js:
```diff
@@ -55,22 +59,30 @@ if (crashReporterDirectory) {
-	// Crashes are stored in the temp directory by default, so we
+	// Crashes are stored in the crashDumps directory by default, so we
 	// need to change that directory to the provided one
 	console.log(`Found --crash-reporter-directory argument. Setting temp directory to be '${crashReporterDirectory}'`);
```
@deepak1556 the console log message talks about `temp`, but it looks like this is `crashDumps` directory now?

bpasero commented on src/main.js:
```diff
@@ -55,22 +59,30 @@ if (crashReporterDirectory) {
+	companyName: companyName,
+	productName: process.env['VSCODE_DEV'] ? `${productName} Dev` : productName,
+	submitURL: submitURL,
+	uploadToServer: !crashReporterDirectory
```
@deepak1556 where is the `extra` metadata carrying version and commit that we had?

deepak1556 replied:
it was never being used by appcenter so I cleaned it up with this refactor.

bpasero commented:
One thing I forgot: we need to offer the user a transition path if the user has configured the crash reporter to be disabled in settings. E.g. we could check this setting on startup and update the `argv.json` automatically + maybe inform the user that a restart is needed.

bpasero commented:
Converting to draft, please ping when ready for final reivew.

deepak1556 commented:
@bpasero I have addressed the feedback, only work left is to accept the crash reporter extra parameters in the appcenter backend. This is ready for review.

bpasero commented on src/vs/workbench/contrib/welcome/telemetryOptOut/browser/telemetryOptOut.ts:
```diff
@@ -139,7 +139,6 @@ export abstract class AbstractTelemetryOptOut implements IWorkbenchContribution
 						run: () => {
 							logTelemetry(true);
 							this.configurationService.updateValue('telemetry.enableTelemetry', false);
```
@deepak1556 this needs to write to the `argv.json` to disable crash reporter, no? How would we otherwise preserve the behaviour we have today?

deepak1556 replied:
this is still pending, how would you suggest we do it ? Is there a way to read the telemetry value on startup ?

bpasero replied:
No, there is not. I would consider to use `IJSONEditingService` for this purpose, it supports comments and allows to preserve the formatting. We use this service to programmatically write into setting files too.

bpasero commented on src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:
```diff
@@ -149,10 +151,12 @@ export class LocalProcessExtensionHost implements IExtensionHost {
 				this._tryListenOnPipe(),
-				this._tryFindDebugPort()
+				this._tryFindDebugPort(),
+				this._telemetryService.getTelemetryInfo()
```
I would certainly not bock the extension host startup to get the telemetry info. We do not seem to wait for the telemetry info in the renderer so I would not wait here unless needed?

bpasero commented on src/vs/workbench/electron-browser/window.ts:
```diff
@@ -418,21 +417,15 @@ export class NativeWindow extends Disposable {
-				this.setupCrashReporter(companyName, productName, this.productService.appCenter, undefined);
-			}
+		// Add extra parameters to crash reporter (if enabled)
+		if (!this.environmentService.disableCrashReporter) {
```
Is this a no-op if the crash reporter is not started? We do only seem to be checking on the CLI wether crash reporter is disabled, but not `argv.json`

bpasero commented on src/bootstrap-fork.js:
```diff
@@ -173,12 +173,27 @@ function terminateWhenParentTerminates() {
-				process['crashReporter'].start(crashReporterOptions);
+			const crashReporterStartOptions = JSON.parse(crashReporterStartOptionsRaw);
+			if (crashReporterStartOptions) {
+				crashReporter.start(crashReporterStartOptions);
```
I am confused, why is this needed? Isn't it enabled for all processes with Electron 9?

bpasero commented on src/vs/workbench/electron-browser/window.ts:
```diff
@@ -418,21 +417,15 @@ export class NativeWindow extends Disposable {
-			}
+		// Add extra parameters to crash reporter (if enabled)
+		if (!this.environmentService.disableCrashReporter) {
+			this.telemetryService.getTelemetryInfo().then(info => {
```
I prefer `async` over `then`, e.g. `(async () => ...)()`

bpasero commented on src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:
```diff
@@ -182,18 +186,32 @@ export class LocalProcessExtensionHost implements IExtensionHost {
-				if (crashesDirectory) {
-					const crashReporterOptions: CrashReporterStartOptions = {
+				// On linux crash reporter needs to be started on child processes explicitly
+				if (platform.isLinux) {
```
So this seems specific to Linux, even with Electron 9 we need to start the crash reporter?

deepak1556 replied:
Only on linux for child node processes the crash reporter has to be started explicitly due to implementation reasons, I have clarified the comment section in the code.

bpasero commented on src/main.js:
```diff
@@ -324,7 +337,10 @@ function createDefaultArgvConfigSync(argvConfigPath) {
-			'	"disable-color-correct-rendering": true'
+			'	"disable-color-correct-rendering": true,',
+			'',
+			'	// Enabled by default to record and send crash reports to Appcenter from all processes',
```
Any particular reason you mention this explicitly here? Since the default is enabled, I wonder why it is called out, for privacy reasons?

deepak1556 replied:
Just trying to be explicit, can remove if not required.

bpasero replied:
Yes I think this is not needed. The other entry makes sense because we explicitly change a default of Electron imho.

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,33 @@ if (crashReporterDirectory) {
+	// * disable-crash-reporter runtime argument is set to 'true'
+	// * --disable-crash-reporter command line parameter is set
+	if (appCenter && (argvConfig['disable-crash-reporter'] === false || argvConfig['disable-crash-reporter'] === 'false') &&
+		!app.commandLine.hasSwitch('disable-crash-reporter')) {
```
Where is this coming from, Chrome?

deepak1556 replied:
No not from chrome but If users start the app with `--disable-crash-reporter`. Isn't that still a supported way to disable it, useful for ci/testing purposes too.

bpasero replied:
@deepak1556 yeah I got confused why you go through `app.commandLine` if all the command args are accessible via the `args` variable already above?

deepak1556 replied:
oh yeah should use that, habitual api call :)

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,33 @@ if (crashReporterDirectory) {
+	if (appCenter && (argvConfig['disable-crash-reporter'] === false || argvConfig['disable-crash-reporter'] === 'false') &&
+		!app.commandLine.hasSwitch('disable-crash-reporter')) {
+		const isWindows = (process.platform === 'win32');
+		const isMacintosh = (process.platform === 'darwin');
```
I see `isMacintosh` is unused variable.

deepak1556 commented:
@bpasero I have the crashreporter starting yet again due to some regressions/complexities faced during testing.

* First idea was to send the appcenter specific data `uid, iid, sid` as extra parameters with `crashReporter.addExtraParameter`. The downside of this api is that the crash report will contain these extra parameters only in the process in which this api is called, that means we have to make sure to call the api in main, all the different renderers and node child processes, which increases the complexity of the implementation.

* Instead I have taken the initial approach to encode all these data in the crash reporter startup which by default will carry these data for all processes. Only thing is that I have eliminated the use of sessionId from telemetryInfo because we are not using it in appcenter before and trying to add that info in the main process will only increase the complexity without any added benefits. And instead of reading from the storage service which uses sqlitedb underneath I have added a simpe one time unique `crashReporterId` that will be stored under the user data dir.

* On linux, for child node processes the crash reporter has to be started explicitly, for which we share the `crashReporterId` from the previous step via command line that again eliminates doing any filesystem operations.

* For disabling the appcenter crash reporter there are now 3 ways
   * Using `--crash-reporter-directory` which will save the crash reports locally
   * Setting `"disable-crash-reporter": true` in `Configure runtime arguments`
   * Using `--disable-crash-reporter` from the command line

Have verified this to work for all 3 platforms and for all the processes we are interested in.

bpasero commented on src/main.js:
```diff
@@ -278,7 +300,8 @@ function readArgvConfigSync() {
 		argvConfig = {
-			'disable-color-correct-rendering': true // Force pre-Chrome-60 color profile handling (for https://github.com/Microsoft/vscode/issues/51791)
+			'disable-color-correct-rendering': true, // Force pre-Chrome-60 color profile handling (for https://github.com/Microsoft/vscode/issues/51791)
+			'disable-crash-reporter': false
```
This is not needed because the default is falsy, no?

bpasero commented on src/bootstrap-fork.js:
```diff
@@ -173,12 +173,13 @@ function terminateWhenParentTerminates() {
 function configureCrashReporter() {
-	const crashReporterOptionsRaw = process.env['CRASH_REPORTER_START_OPTIONS'];
-	if (typeof crashReporterOptionsRaw === 'string') {
+	const crashReporterStartOptionsRaw = process.env['CRASH_REPORTER_START_OPTIONS'];
```
All the changes in here seem to be just renames, so for the sake of a smaller PR, I would undo them.

bpasero commented on src/main.js:
```diff
@@ -358,6 +384,18 @@ function getArgvConfigPath() {
 	return path.join(os.homedir(), dataFolderName, 'argv.json');
 }
 
+async function getCrashReporterId() {
```
I am not a big fan of having to read another file on startup. As I suggested initially, why not simply write this into `argv.json`? E.g. we could have a crash reporter related key in that file that is:
* `false` when disabled
* `UUID` when enabled and used as crash reporter ID
* `undefined` when we first startup with these changes and need to write a new UUID into

deepak1556 replied:
there is now `enable-crash-reporter` which will take those 3 states and be written to argv.json

* `"enable-crash-reporter": false` which users can write
* `"enable-crash-reporter": <UUID>` when the key is undefined we create a UUID and write to the file, otherwise the value will be used as crash reporter id for any future calls.

bpasero commented on src/main.js:
```diff
@@ -31,12 +36,11 @@ const args = parseCLIArgs();
+// If a crash-reporter-directory is specified we store the crash reports
+// in the specified directory and don't upload them to the crash server.
 let crashReporterDirectory = args['crash-reporter-directory'];
 if (crashReporterDirectory) {
```
This entire `if-else` block would probably read better if:
* the crash reporter ID comes from `argv.json` (as discussed below) so you avoid a promise
* the `crashReporter.start` method is only called once and the properties are given depending on the command line args instead of starting it from each if-else branch

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,42 @@ if (crashReporterDirectory) {
+crashReporter.start({
+	companyName: companyName,
+	productName: process.env['VSCODE_DEV'] ? `${productName} Dev` : productName,
+	submitURL: submitURL,
```
```suggestion
	submitURL,
```

bpasero commented on src/vs/workbench/contrib/welcome/telemetryOptOut/browser/telemetryOptOut.ts:
```diff
@@ -136,9 +140,10 @@ export abstract class AbstractTelemetryOptOut implements IWorkbenchContribution
 							logTelemetry(true);
 							this.configurationService.updateValue('telemetry.enableTelemetry', false);
+							await this.jsonEditingService.write(this.environmentService.argvResource, [{ path: ['enable-crash-reporter'], value: false }], true);
 							this.configurationService.updateValue('telemetry.enableCrashReporter', false);
```
@deepak1556 the entire `telemetry.enableCrashReporter` config should go away, no? Otherwise a user may assume that changing `telemetry.enableCrashReporter` disables the crash reporter?

bpasero commented on src/vs/workbench/electron-browser/window.ts:
```diff
@@ -415,25 +416,17 @@ export class NativeWindow extends Disposable {
 		});
 
+		// Edit argv.json with crash reporter configs for a fresh start.
+		if (this.environmentService.createCrashReporterConfig) {
```
@deepak1556 I would actually move this into `electron-main` because if a user opens multiple windows at once, they might all fight over this in parallel. 

Since `IConfigurationService` can be accessed in `electron-main` too, [`afterWindowOpen`](https://github.com/microsoft/vscode/blob/bbe787ac280cab75002b71460a93ab725d810e80/src/vs/code/electron-main/app.ts#L805) would be a good fit imho since this can happen after windows opened.

One downside is that I do not see the json editing service available in `electron-main`, but the primitives for writing should be available: https://github.com/microsoft/vscode/blob/bbe787ac280cab75002b71460a93ab725d810e80/src/vs/base/common/jsonEdit.ts#L1

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,42 @@ if (crashReporterDirectory) {
+	// * --crash-reporter-directory is specified
+	// * enable-crash-reporter runtime argument is set to 'false'
+	// * --disable-crash-reporter command line parameter is set
+	if (appCenter && (argvConfig['enable-crash-reporter'] !== false) && !args['disable-crash-reporter']) {
```
@deepak1556 isn't this starting the crash reporter even if this is the first startup where `argvConfig['enable-crash-reporter'] === undefined`? Don't we need to wait until we have a crash reporter ID for this to work?

deepak1556 replied:
oh yeah should fix that, but without crash reporter ID it will be a noop even if called on the server side. But avoiding the call is all the more better.

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,42 @@ if (crashReporterDirectory) {
 }
 
+// Start crash reporter for all processes
+crashReporter.start({
```
I wonder how useful it is to start the crash reporter if the user has disabled it via settings?

deepak1556 replied:
There is not harm in calling it irrespective of the settings, when the appcenter config is not set, it won't send any data to the server, adding this conditionally will only make this `if else` block less simple with no added benefits.

bpasero replied:
Ok, just wanted to confirm.

deepak1556 commented on src/vs/workbench/services/extensions/electron-browser/localProcessExtensionHost.ts:
```diff
@@ -182,18 +182,23 @@ export class LocalProcessExtensionHost implements IExtensionHost {
+						crashReporterStartOptions.submitURL = submitURL.concat('&uid=', crashReporterId, '&iid=', crashReporterId, '&sid=', crashReporterId);
+						crashReporterStartOptions.uploadToServer = true;
+					}
+					opts.env.CRASH_REPORTER_START_OPTIONS = JSON.stringify(crashReporterStartOptions);
```
Its not always sent, `crashReporterId` is only available when app center crash reporting is enabled in the main process.

bpasero replied:
@deepak1556 but if `CRASH_REPORTER_START_OPTIONS` is defined, we will always end up starting the crash reporter here:

https://github.com/microsoft/vscode/blob/3825520a4573913a1b06de486a1b41fc26968658/src/bootstrap-fork.js#L177-L186

deepak1556 replied:
the behavior would be same as https://github.com/microsoft/vscode/pull/102011#discussion_r453737406 from the main process.

deepak1556 replied:
Aim here in general is to avoid setting any appcenter config when crash reporter is disabled by user.

bpasero replied:
Yeah makes sense.

deepak1556 commented on src/vs/code/electron-main/app.ts:
```diff
@@ -815,6 +819,29 @@ export class CodeApplication extends Disposable {
+		const argvString = argvContent.value.toString();
+		const argvJSON = JSON.parse(stripComments(argvString));
+		if (argvJSON['enable-crash-reporter'] === undefined) {
+			const enableCrashReporter = this.configurationService.getValue<boolean>('telemetry.enableCrashReporter');
```
this returns `undefined` in the main process. Is it too early to check the value ?

bpasero replied:
@deepak1556 configuration service in `electron-main` has no defaults because they get registered in renderer only. @sandy081 might have an issue for that, but for now we have to fill in the default value manually.

sandy081 replied:
Yeah this is the limitation of using configuration service in non renderer process (main, shared). There could be an issue exists already but not sure.

bpasero commented on src/vs/workbench/electron-sandbox/desktop.contribution.ts:
```diff
@@ -344,6 +344,14 @@ import { IJSONSchema } from 'vs/base/common/jsonSchema';
 			},
+			'enable-crash-reporter': {
+				type: 'boolean',
+				markdownDescription: nls.localize('argv.enableCrashReporter', 'Allows disabling crash reporting to Appcenter, should restart the app if the value is changed.')
```
@deepak1556 I wonder if we need to mention `AppCenter` here, since this may change over time. I would change it to be more abstract and talk about crash reporting but not the endpoint.

bpasero commented on src/vs/platform/environment/node/argv.ts:
```diff
@@ -62,8 +62,8 @@ export interface ParsedArgs {
 	'export-default-configuration'?: string;
 	'install-source'?: string;
 	'disable-updates'?: boolean;
-	'disable-crash-reporter'?: boolean;
```
@deepak1556 should we keep `disable-crash-reporter` in `argv.ts` just because we read it from `main.js` still? I think it helps document the list of CLI flags we support, even if only used in one place.

bpasero commented on src/main.js:
```diff
@@ -55,23 +61,36 @@ if (crashReporterDirectory) {
+	if (appCenter && argvConfig['enable-crash-reporter'] && !args['disable-crash-reporter']) {
+		const isWindows = (process.platform === 'win32');
+		const isLinux = (process.platform === 'linux');
+		const crashReporterId = argvConfig['crash-reporter-id'];
```
@deepak1556 should we validate here that the `crashReporterId` is a `UUID`? After all, the user may have tampered with the value and we send this back to us.

deepak1556 commented:
@bpasero should be good to go, PTAL.

deepak1556 commented:
Internal builds with working devtools have been updated in `vscode-electron-prebuilt`, this PR is ready to be landed.

bpasero commented:
@deepak1556 let me do the `app.focus` change we talked about and then merge this in?

deepak1556 commented:
sounds good, feel free to merge the PR at your convenience. I am gonna sign-off for the day :)

bpasero commented on remote/.yarnrc:
```diff
@@ -1,3 +1,3 @@
 disturl "http://nodejs.org/dist"
-target "12.4.0"
+target "12.14.1"
```
@alexdima fyi not sure why `remote` locked the dependency to `12.4.0` but it would now advance to the same as Electron 9

alexdima replied:
:+1: Maybe there was no node release with the same version number as the one used in Electron?
